examples/filesaver/main.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	lksdk "github.com/livekit/server-sdk-go"
+	"github.com/livekit/server-sdk-go/pkg/samplebuilder"
+	"github.com/pion/rtp/codecs"
+	"github.com/pion/webrtc/v3"
+	"github.com/pion/webrtc/v3/pkg/media"
+	"github.com/pion/webrtc/v3/pkg/media/h264writer"
+	"github.com/pion/webrtc/v3/pkg/media/ivfwriter"
+	"github.com/pion/webrtc/v3/pkg/media/oggwriter"
+)
+
+var (
+	host, apiKey, apiSecret, roomName, identity string
+)
+
+func init() {
+	flag.StringVar(&host, "host", "", "livekit server host")
+	flag.StringVar(&apiKey, "api-key", "", "livekit api key")
+	flag.StringVar(&apiSecret, "api-secret", "", "livekit api secret")
+	flag.StringVar(&roomName, "room-name", "", "room name")
+	flag.StringVar(&identity, "identity", "", "participant identity")
+}
+
+func main() {
+	flag.Parse()
+	if host == "" || apiKey == "" || apiSecret == "" || roomName == "" || identity == "" {
+		fmt.Println("invalid arguments.")
+		return
+	}
+	room, err := lksdk.ConnectToRoom(host, lksdk.ConnectInfo{
+		APIKey:              apiKey,
+		APISecret:           apiSecret,
+		RoomName:            roomName,
+		ParticipantIdentity: identity,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	room.Callback.OnTrackSubscribed = trackSubscribed
+	// room.Callback.OnTrackUnsubscribed = nil
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT)
+
+	<-sigChan
+	room.Disconnect()
+}
+
+func trackSubscribed(track *webrtc.TrackRemote, publication *lksdk.RemoteTrackPublication, rp *lksdk.RemoteParticipant) {
+	fileName := fmt.Sprintf("%s-%s", rp.Identity(), track.ID())
+	fmt.Println("write track to file ", fileName)
+	NewTrackWriter(track, rp.WritePLI, fileName)
+}
+
+const (
+	maxLate = 200
+)
+
+type TrackWriter struct {
+	sb     *samplebuilder.SampleBuilder
+	writer media.Writer
+	track  *webrtc.TrackRemote
+}
+
+func NewTrackWriter(track *webrtc.TrackRemote, pliWriter lksdk.PLIWriter, fileName string) (*TrackWriter, error) {
+	var (
+		sb     *samplebuilder.SampleBuilder
+		writer media.Writer
+		err    error
+	)
+	switch {
+	case strings.EqualFold(track.Codec().MimeType, "video/vp8"):
+		sb = samplebuilder.New(maxLate, &codecs.VP8Packet{}, track.Codec().ClockRate, samplebuilder.WithPacketDroppedHandler(func() {
+			pliWriter(track.SSRC())
```

Comment:
in addition to packets being dropped, we should write it periodically to ensure keyframes are generated every N seconds, for seekability

Reply:
for playback enhancement(seekability, reduce file size and others), we'd better to transcode after record completed. send pli periodically will affect other participant in the room (higher bitrate may cause higher packet loss)